File: go/test/economic_development_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/las-vegas-city-sdk/go"
	"github.com/voxgig-sdk/las-vegas-city-sdk/go/core"

	vs "github.com/voxgig-sdk/las-vegas-city-sdk/go/utility/struct"
)

func TestEconomicDevelopmentEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EconomicDevelopment(nil)
		if ent == nil {
			t.Fatal("expected non-nil EconomicDevelopmentEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"economic_development": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.EconomicDevelopment(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.EconomicDevelopment(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := economic_developmentBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "economic_development." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LAS_VEGAS_CITY_TEST_ECONOMIC_DEVELOPMENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		economicDevelopmentRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.economic_development", setup.data)))
		var economicDevelopmentRef01Data map[string]any
		if len(economicDevelopmentRef01DataRaw) > 0 {
			economicDevelopmentRef01Data = core.ToMapAny(economicDevelopmentRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = economicDevelopmentRef01Data

		// LIST
		economicDevelopmentRef01Ent := client.EconomicDevelopment(nil)
		economicDevelopmentRef01Match := map[string]any{}

		economicDevelopmentRef01ListResult, err := economicDevelopmentRef01Ent.List(economicDevelopmentRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, economicDevelopmentRef01ListOk := economicDevelopmentRef01ListResult.([]any)
		if !economicDevelopmentRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", economicDevelopmentRef01ListResult)
		}

	})
}

func economic_developmentBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "economic_development", "EconomicDevelopmentTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read economic_development test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse economic_development test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"economic_development01", "economic_development02", "economic_development03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LAS_VEGAS_CITY_TEST_ECONOMIC_DEVELOPMENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LAS_VEGAS_CITY_TEST_ECONOMIC_DEVELOPMENT_ENTID": idmap,
		"LAS_VEGAS_CITY_TEST_LIVE":      "FALSE",
		"LAS_VEGAS_CITY_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["LAS_VEGAS_CITY_TEST_ECONOMIC_DEVELOPMENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LAS_VEGAS_CITY_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewLasVegasCitySDK(core.ToMapAny(mergedOpts))
	}

	live := env["LAS_VEGAS_CITY_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LAS_VEGAS_CITY_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/EconomicDevelopmentEntityTest.php
<?php
declare(strict_types=1);

// EconomicDevelopment entity test

require_once __DIR__ . '/../lasvegascity_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EconomicDevelopmentEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = LasVegasCitySDK::test(null, null);
        $ent = $testsdk->EconomicDevelopment(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = economic_development_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "economic_development." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set LASVEGASCITY_TEST_ECONOMIC_DEVELOPMENT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $economic_development_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.economic_development")));
        $economic_development_ref01_data = null;
        if (count($economic_development_ref01_data_raw) > 0) {
            $economic_development_ref01_data = Helpers::to_map($economic_development_ref01_data_raw[0][1]);
        }

        // LIST
        $economic_development_ref01_ent = $client->EconomicDevelopment(null);
        $economic_development_ref01_match = [];

        [$economic_development_ref01_list_result, $err] = $economic_development_ref01_ent->list($economic_development_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($economic_development_ref01_list_result);

    }
}

function economic_development_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/economic_development/EconomicDevelopmentTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = LasVegasCitySDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["economic_development01", "economic_development02", "economic_development03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("LASVEGASCITY_TEST_ECONOMIC_DEVELOPMENT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "LASVEGASCITY_TEST_ECONOMIC_DEVELOPMENT_ENTID" => $idmap,
        "LASVEGASCITY_TEST_LIVE" => "FALSE",
        "LASVEGASCITY_TEST_EXPLAIN" => "FALSE",
        "LASVEGASCITY_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["LASVEGASCITY_TEST_ECONOMIC_DEVELOPMENT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["LASVEGASCITY_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["LASVEGASCITY_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new LasVegasCitySDK(Helpers::to_map($merged_opts));
    }

    $live = $env["LASVEGASCITY_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["LASVEGASCITY_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
